includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -2785,6 +2785,11 @@ public function capture_charge( $order, $include_level3 = true, $intent_metadata
 			try {
 				$error_message = $e->getMessage();
 				$http_code     = $e->get_http_code();
+				$error_code    = $e->get_error_code();
+
+				if ( 'amount_too_large' === $error_code ) {
```

Comment:
That's a simple change, but I have mixed feelings about where it has been placed.
Are there any alternatives where this message normalization can be placed?

Reply:
Indeed @kalessil, there's `generate_capture_failed_note` method, seems a more suitable place.  https://github.com/Automattic/woocommerce-payments/blob/ea5370017345fc2254f5011e4fdf381b3842e18a/includes/class-wc-payments-order-service.php#L1142

This method is called in `mark_payment_capture_failed` https://github.com/Automattic/woocommerce-payments/blob/ea5370017345fc2254f5011e4fdf381b3842e18a/includes/class-wc-payments-order-service.php#L204

Reply:
There is also one more opportunity: https://github.com/Automattic/woocommerce-payments/blob/develop/includes/wc-payment-api/class-wc-payments-api-client.php#L1914

Can you compare 2 of them weighing the pros and cons a comment?

Reply:
> Can you compare 2 of them weighing the pros and cons a comment?

Sure! But before getting into the details please note that I already explored one option by the time I noticed your comment (https://github.com/Automattic/woocommerce-payments/pull/7581/commits/5630836f515770995ae422620efb36a816350b46) so I explored your proposal in an additional commit https://github.com/Automattic/woocommerce-payments/pull/7581/commits/d8bbe34c08708bd64c46c2c09e8c376a04b30901

### A. Modify order service (https://github.com/Automattic/woocommerce-payments/pull/7581/commits/5630836f515770995ae422620efb36a816350b46)

**Pros**

- It overrides the message in the context of capturing a charge, where the error may happen.

**Cons**

- I'm passing the `error_code` from the gateway down to the `generate_capture_failed_note` in the order service, going through `mark_payment_capture_failed`. This method does nothing with this parameter, only passes it to `generate_capture_failed_note`.

### B. Introduce Amount_Too_Large_Exception (https://github.com/Automattic/woocommerce-payments/pull/7581/commits/d8bbe34c08708bd64c46c2c09e8c376a04b30901)

**Pros**
- Centralized solution.
- Does not require changes to the current flow of capturing a charge

**Cons**
- If I am not wrong, the `amount_too_large` error is not specific to capturing charges. I'm checking that the payment intent status  is `requires_capture`  before throwing the `Amount_Too_Large_Exception`. But it feels like dealing with a specific issue of the capture flow in a more generic place such as the API client.

I'm leaning towards option B, as it seems a cleaner solution to me. What do you think?

Reply:
Well, the whole point of asking to compare the options is to double-check if you notice that option A would require a change in public method `mark_payment_capture_failed` (passing in an error code along with the message).

The problem with public methods is that they might be consumed already by in-house or custom merchant integrations.  This means it's a no-go from the beginning (for the draft, probably ok, but still a point to concern).

Option B is an isolated solution, which is flexible so yes, I think it's the best shot for our needs.